### DISCUSSION TITLE
Use session.add to insert an ORM instance cascading all relationships

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#8012] [Usage Questions] [answered]
I have a model with multiple relationships

```python
class Model1(Base):
    __tablename__ = 'model1'
    id = Column(Integer, primary_key=True)
    name =  Column(String(50))
    model2_id = Column (Integer, ForeignKey('model2.id'))
    rel_2 = relationship("Model2")

class Model2(Base):
    __tablename__ = 'model2'
    id = Column(Integer, primary_key=True)
    name =  Column(String(50))
    model3_id = Column (Integer,   ForeignKey('model3.id'))
    rel_3 = relationship("Model3")

class Model3(Base):
    __tablename__ = 'model3'
    id = Column(Integer, primary_key=True)
    name =  Column(String(50))    
```

I try to create manually an instance of Model1 by setting relationships from client inputs.

```python
model_3_instance = Model3(name="mymodel3", id=3)
# here I set the id manually as it is a part of the client input !

model_2_instance = Model2(name="mymodel2" rel_3=[model_3_instance])

model_1_instance = Model1(name="mymodel1" rel_2=[model_2_instance])

session.add(model_1_instance)
session.commit()
```
And everything is working perfectly ! All the rows of each realtionships are added automatically just by adding the model_1 instance to the session and commit it.
BUT : for Model3, the id is a client input. So if it was already added by the past, it will throw an Integrity error ... Because it will try to insert something that already exists with the same id.

What I need is a kind of get or create or an insert with IGNORE, but I need to perform thos this using the convenience of the `session.add()` method that will add the relationships automatically. 

Is it possible ?

**Comment by CaselIT** [comment, marked as answer]
Hi, 

I think not directly. You could try `session.merge(model_1_instance)` but I don't think the merge will cascade to the relationship.
If it does not cascade then you could do this
```py
model_3_instance = Model3(name="mymodel3", id=3)
# here I set the id manually as it is a part of the client input !
model_3_instance = session.merge(model_3_instance) 
model_2_instance = Model2(name="mymodel2" rel_3=[model_3_instance])

model_1_instance = Model1(name="mymodel1" rel_2=[model_2_instance])

session.add(model_1_instance)
session.commit()
```

This will merge the model_3 to the session, so that if id=3 already exists the instance from the db is used, with the updates passed by the user

**Reply by jossefaz** [reply]
`session.merge(model_1_instance)` worked perfectly in my case...and it also cascades the relationships.

**Comment by zzzeek** [comment]
if the object refers to a primary key that already exists, you need to load the existing object from that id that's been passed by the client.

Using a merge or a recipe such as [UniqueObject](https://github.com/sqlalchemy/sqlalchemy/wiki/UniqueObject) are ways to maybe make it look more streamlined but IMO if the client is just sending you "id", load up that object if not exists.
